src/uct/tcp/tcp_ep.c

```diff
@@ -329,139 +367,265 @@ void uct_tcp_ep_set_failed(uct_tcp_ep_t *ep)
+    struct sockaddr_in dest_addr;
     ucs_status_t status;
 
-    status = uct_tcp_ep_create_socket_and_connect(iface, dest_addr,
-                                                  conn_sn, ep_p);
+    if (ucs_test_all_flags(params->field_mask,
+                           UCT_EP_PARAM_FIELD_DEV_ADDR |
+                           UCT_EP_PARAM_FIELD_IFACE_ADDR)) {
+        uct_tcp_ep_set_dest_addr(params->dev_addr, params->iface_addr,
+                                 &dest_addr);
+        ep_dest_addr = &dest_addr;
+    }
+
+    status = uct_tcp_ep_init(iface, -1, ep_dest_addr, &ep);
     if (status != UCS_OK) {
         return status;
     }
 
-    uct_tcp_ep_add_ctx_cap(*ep_p, UCT_TCP_EP_FLAG_CTX_TYPE_TX);
+    uct_tcp_cm_ep_set_conn_sn(ep);
 
-    if (!uct_tcp_ep_is_self(*ep_p)) {
-        uct_tcp_iface_remove_ep(*ep_p);
-        uct_tcp_cm_insert_ep(iface, *ep_p);
+    if (!(ep->flags & UCT_TCP_EP_FLAG_CONNECT_TO_EP)) {
+        status = uct_tcp_ep_connect(ep);
+        if (status != UCS_OK) {
+            return status;
+        }
     }
 
+    /* cppcheck-suppress autoVariables */
+    *ep_p = &ep->super.super;
     return UCS_OK;
 }
 
-ucs_status_t uct_tcp_ep_create(const uct_ep_params_t *params,
-                               uct_ep_h *ep_p)
+ucs_status_t uct_tcp_ep_get_address(uct_ep_h tl_ep, uct_ep_addr_t *ep_addr)
 {
-    uct_tcp_iface_t *iface = ucs_derived_of(params->iface, uct_tcp_iface_t);
-    uct_tcp_ep_t *ep       = NULL;
-    struct sockaddr_in dest_addr;
-    uct_tcp_cm_conn_sn_t conn_sn;
-    ucs_status_t status;
+    uct_tcp_ep_t *ep        = ucs_derived_of(tl_ep, uct_tcp_ep_t);
+    uct_tcp_ep_addr_t *addr = (uct_tcp_ep_addr_t*)ep_addr;
 
-    UCT_EP_PARAMS_CHECK_DEV_IFACE_ADDRS(params);
-    memset(&dest_addr, 0, sizeof(dest_addr));
-    /* TODO: handle AF_INET6 */
-    dest_addr.sin_family = AF_INET;
-    dest_addr.sin_port   = *(in_port_t*)params->iface_addr;
-    dest_addr.sin_addr   = *(const struct in_addr*)ucs_sockaddr_get_inet_addr
-                                                   ((struct sockaddr*)params->dev_addr);
+    ucs_assert(ep->flags & UCT_TCP_EP_FLAG_CONNECT_TO_EP);
 
-    conn_sn = uct_tcp_cm_get_conn_sn(iface, &dest_addr);
+    addr->conn_sn = ep->conn_sn;
+    return uct_iface_get_address(tl_ep->iface,
+                                 (uct_iface_addr_t*)&addr->iface_addr);
+}
 
-    if (uct_tcp_iface_is_self_addr(iface, &dest_addr)) {
-        goto out_create_ep;
+ucs_status_t uct_tcp_ep_connect_to_ep(uct_ep_h tl_ep,
+                                      const uct_device_addr_t *dev_addr,
+                                      const uct_ep_addr_t *ep_addr)
+{
+    uct_tcp_ep_t *ep        = ucs_derived_of(tl_ep, uct_tcp_ep_t);
+    uct_tcp_iface_t *iface  = ucs_derived_of(ep->super.super.iface,
+                                             uct_tcp_iface_t);
+    uct_tcp_ep_addr_t *addr = (uct_tcp_ep_addr_t*)ep_addr;
+    uct_tcp_ep_t *peer_ep;
+    int connect_to_self;
```

Comment:
minor: can init in place

Reply:
done

Reply:
not really

Reply:
sorry, forgot to say that it was not applied
added comment at 663
`uct_tcp_ep_set_dest_addr()` has to be called prior to set the address